src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs

```diff
@@ -79,7 +79,9 @@ public async Task Invoke(HttpContext context)
     {
         var options = _options.CurrentValue;
 
-        var elements = new string[_fieldsLength];
+        var additionalHeadersLength = _options.CurrentValue.AdditionalRequestHeaders?.Count ?? 0;
+
+        var elements = new string[_fieldsLength + additionalHeadersLength];
```

Comment:
We should do something to ensure that we clone the set of request headers once we start using it. And we also need to make sure the same cloned set is used in `W3CLoggerProcessor`. Just because it's an `IReadOnlyList<string>` to us doesn't mean it's not an easily mutable `List<string>` to whoever assigned it.

If someone added a header at a bad time, the W3CLoggingMiddleware could throw an `IndexOutOfRangeException`. And even if we avoided that, the headers for the W3C log file would be wrong.

Reply:
This reminds me, `FileLoggerProcessor` moves to a new file when the `LoggingFields` change: https://github.com/dotnet/aspnetcore/blob/5ff6e316a8e34bb06b986ec13ec7aae3f61e886b/src/Middleware/HttpLogging/src/FileLoggerProcessor.cs#L64-L74

This PR should do the same when `AdditionalRequestHeaders` changes

Reply:
The headers list is now cloned.